sklearn/utils/_array_api.py

```diff
@@ -822,6 +823,74 @@ def _estimator_with_converted_arrays(estimator, converter):
     return new_estimator
 
 
+def make_converter(X):
+    """Helper to implement the 'y follows X' rule.
+
+    Returns a function that converts an array to the namespace and device of X.
+
+    When X is not an array api array, the converter does nothing.
+    """
+    xp, is_array_api = get_namespace(X)
+    if not is_array_api:
+
+        def convert(data):
+            return data
+
+        return convert
+    else:
+        device_ = device(X)
+
+        def convert(data):
+            if data is None or isinstance(data, (numbers.Number, str)):
+                return data
+            data_xp, data_is_array_api, data_device = get_namespace_and_device(data)
+            if data_xp == xp and data_device == device_:
+                return data
+            if not data_is_array_api:
+                return xp.asarray(data, device=device_)
+            try:
+                return xp.asarray(data, device=device_)
+            except Exception:
+                # direct conversion to a different library may fail in which
+                # case we try converting to numpy first
+                data = _convert_to_numpy(data, data_xp)
+                return xp.asarray(data, device=device_)
+
+        return convert
+
+
+def convert_attributes(estimator, ref_array):
+    return _estimator_with_converted_arrays(estimator, make_converter(ref_array))
+
+
+def check_fitted_attribute(estimator, method_name, attr_name, X):
```

Comment:
during the meeting we said that this check could be added to `utils.validation.check_is_fitted`, but at some point we also said that array API support shouldn't modify scikit-learn's public API yet (which `check_is_fitted` is part of)